### PR TITLE
 [various] clean stat buttons

### PR DESCRIPTION
**Pull request by ppr-odoo**
Pad:https://pad.odoo.com/p/r.a539473706deabca72f72e98f60b751f
Task::https://www.odoo.com/web?debug=assets#id=2024467&action=333&active_id=967&model=project.task&view_type=form&menu_id=4720

### CODE REVIEW

File: addons/repair/models/repair.py [hunk 1]
```diff
@@ -115,6 +115,12 @@ def _default_stock_location(self):
     amount_tax = fields.Float('Taxes', compute='_amount_tax', store=True)
     amount_total = fields.Float('Total', compute='_amount_total', store=True)
     tracking = fields.Selection('Product Tracking', related="product_id.tracking", readonly=False)
+    invoice_count = fields.Integer(string='Invoice Count', compute='_get_invoiced')
+
+    @api.depends('invoice_id')
+    def _get_invoiced(self):
+        for repair in self:
+            repair.invoice_count = len(repair.invoice_id)
```
Comment (sswapnesh) [inline addons/repair/models/repair.py]: `invoice_id` is m2o so Do we need `invoice_count` in this case ?
(or Did I missed something ?)

File: addons/sale/models/sale.py [hunk 2]
```diff
@@ -498,6 +498,14 @@ def action_view_invoice(self):
             action['res_id'] = invoices.id
         else:
             action = {'type': 'ir.actions.act_window_close'}
+        action['context'] = {
+            'default_type': 'out_invoice',
+            'default_partner_id': self.partner_id.id,
+            'default_partner_shipping_id': self.partner_shipping_id.id,
+            'default_invoice_payment_term_id': self.payment_term_id.id,
+            'default_invoice_origin': self.name,
+            'default_user_id': self.user_id.id,
```
Comment (sswapnesh) [inline addons/sale/models/sale.py]: This will be always from `default_get`

File: addons/sale_crm/models/crm_lead.py [hunk 3]
```diff
@@ -83,3 +83,29 @@ def action_new_quotation(self):
             'default_source_id': self.source_id.id
         }
         return action
+
+    @api.multi
+    def action_sale_crm_order(self):
```
Comment (sswapnesh) [inline addons/sale_crm/models/crm_lead.py]:
```suggestion
    def action_view_sale_order(self):
```

File: addons/sale_crm/models/crm_lead.py [hunk 4]
```diff
@@ -83,3 +83,29 @@ def action_new_quotation(self):
...
+        '''
+        This function returns an action that display existing sale quotations/order
+        of given sales quotations/order ids. It can either be a in a list or in a form
```
Comment (sswapnesh) [inline addons/sale_crm/models/crm_lead.py]: quotations/order --> quotation/order

File: addons/sale_crm/models/crm_lead.py [hunk 5]
```diff
@@ -83,3 +83,29 @@ def action_new_quotation(self):
...
+        view, if there is only one sale quotations/order to show.
+        '''
+        order_type = self.env.context.get('order_type')
+        ctx = self._context.copy()
+        ctx.update({
+            'search_default_partner_id': self.partner_id.id,
+            'default_partner_id': self.partner_id.id,
+        })
+        if(order_type == "quotation"):
```
Comment (sswapnesh) [inline addons/sale_crm/models/crm_lead.py]:
```suggestion
        if order_type == "quotation":
```
Maybe

File: addons/sale_crm/models/crm_lead.py [hunk 6]
```diff
@@ -83,3 +83,29 @@ def action_new_quotation(self):
...
+            action = self.env.ref('sale_crm.sale_action_quotations').read()[0]
+            ctx.update({'search_default_draft': 1, 'search_default_sent': 1})
+        else:
+            action = self.env.ref('sale_crm.sale_action_orders').read()[0]
+            ctx.update({'search_default_sale': 1})
+        orders = self.mapped('order_ids')
+        if len(orders) <= 1:
```
Comment (sswapnesh) [inline addons/sale_crm/models/crm_lead.py]:
```suggestion
        if len(orders) == 1:
```

File: addons/sale_quotation_builder/views/sale_order_template_views.xml [hunk 7]
```diff
@@ -7,9 +7,10 @@
         <field name="model">sale.order.template</field>
         <field name="type">form</field>
         <field name="arch" type="xml">
-
-            <xpath expr="//div[@name='button_box']" position="inside">
-                <button string="Edit Template" type="object" name="open_template" class="oe_stat_button" icon="fa-pencil"/>
+            <xpath expr="//sheet" position='before'>
+                <header>
+                    <button string="Edit Template" type="object" name="open_template" class="btn-primary"/>
```
Comment (sswapnesh) [inline addons/sale_quotation_builder/views/sale_order_template_views.xml]: Good to align string and method name

File: addons/sale_stock/models/sale_order.py [hunk 8]
```diff
@@ -137,6 +138,7 @@ def action_view_delivery(self):
         elif pickings:
             action['views'] = [(self.env.ref('stock.view_picking_form').id, 'form')]
             action['res_id'] = pickings.id
+        action['context'] = dict(self._context, default_partner_id=self.partner_id.id, default_picking_type_id=pickings[0].picking_type_id.id, default_sale_id=self.id, default_origin=self.name)
```
Comment (sswapnesh) [inline addons/sale_stock/models/sale_order.py]: Add `self.ensure_one()` as you are already touching this

**Comment by tivisse** [comment]
@robodoo r+

**Comment by mart-e** [comment]
> See https://www.odoo.com/web#id=2024467&action=333&active_id=1251&model=project.task&view_type=form&menu_id=4720 for the detailed list of all the requested improvements.

Not a big deal but could you use the URL of the read-only address of the pad instead in the commit message?
It is public (and, bonus, NGR does not like that)

File: addons/crm_iap_lead_website/models/crm_reveal_rule.py [hunk 9]
```diff
@@ -112,11 +112,13 @@ def _compute_leads_count(self):
     def action_get_lead_tree_view(self):
         action = self.env.ref('crm.crm_lead_all_leads').read()[0]
         action['domain'] = [('id', 'in', self.lead_ids.ids), ('type', '=', 'lead')]
+        action['context'] = dict(self._context, create=False)
```
Comment (Julien00859) [inline addons/crm_iap_lead_website/models/crm_reveal_rule.py:115]: Context leaking

File: addons/crm_iap_lead_website/models/crm_reveal_rule.py [hunk 10]
```diff
@@ -112,11 +112,13 @@ def _compute_leads_count(self):
...
         return action
 
     def action_get_opportunity_tree_view(self):
         action = self.env.ref('crm.crm_lead_opportunities').read()[0]
         action['domain'] = [('id', 'in', self.lead_ids.ids), ('type', '=', 'opportunity')]
+        action['context'] = dict(self._context, create=False)
```
Comment (Julien00859) [inline addons/crm_iap_lead_website/models/crm_reveal_rule.py:121]: Context leaking

File: addons/link_tracker/models/link_tracker.py [hunk 11]
```diff
@@ -167,6 +167,7 @@ def convert_links(self, html, vals, blacklist=None):
     def action_view_statistics(self):
         action = self.env['ir.actions.act_window'].for_xml_id('link_tracker', 'link_tracker_click_action_statistics')
         action['domain'] = [('link_id', '=', self.id)]
+        action['context'] = dict(self._context, create=False)
```
Comment (Julien00859) [inline addons/link_tracker/models/link_tracker.py:170]: Context leaking

File: addons/mrp/models/mrp_production.py [hunk 12]
```diff
@@ -246,6 +246,7 @@ def action_view_mo_delivery(self):
         elif pickings:
             action['views'] = [(self.env.ref('stock.view_picking_form').id, 'form')]
             action['res_id'] = pickings.id
+        action['context'] = dict(self._context, default_origin=self.name, create=False)
```
Comment (Julien00859) [inline addons/mrp/models/mrp_production.py:249]: Context leaking

File: addons/mrp/models/mrp_production.py [hunk 13]
```diff
@@ -916,6 +917,7 @@ def action_see_move_scrap(self):
         self.ensure_one()
         action = self.env.ref('stock.action_stock_scrap').read()[0]
         action['domain'] = [('production_id', '=', self.id)]
+        action['context'] = dict(self._context, default_origin=self.name)
```
Comment (Julien00859) [inline addons/mrp/models/mrp_production.py:920]: Context leaking

File: addons/purchase_stock/models/stock.py [hunk 14]
```diff
@@ -205,4 +205,5 @@ def action_view_po(self):
         self.ensure_one()
         action = self.env.ref('purchase.purchase_form_action').read()[0]
         action['domain'] = [('id', 'in', self.mapped('purchase_order_ids.id'))]
+        action['context'] = dict(self._context, create=False)
```
Comment (Julien00859) [inline addons/purchase_stock/models/stock.py:208]: Context leaking

File: addons/sale_stock/models/sale_order.py [hunk 15]
```diff
@@ -134,6 +134,8 @@ def action_view_delivery(self):
         elif pickings:
             action['views'] = [(self.env.ref('stock.view_picking_form').id, 'form')]
             action['res_id'] = pickings.id
+        picking_id = pickings.filtered(lambda l: l.picking_type_id.code == 'outgoing')[0]
+        action['context'] = dict(self._context, default_partner_id=self.partner_id.id, default_picking_id=picking_id.id, default_picking_type_id=picking_id.picking_type_id.id, default_origin=self.name, default_group_id=picking_id.group_id.id)
```
Comment (Julien00859) [inline addons/sale_stock/models/sale_order.py:138]: Context leaking

File: addons/sale_stock/models/stock.py [hunk 16]
```diff
@@ -132,4 +132,5 @@ def action_view_so(self):
         self.ensure_one()
         action = self.env.ref('sale.action_orders').read()[0]
         action['domain'] = [('id', 'in', self.mapped('sale_order_ids.id'))]
+        action['context'] = dict(self._context, create=False)
```
Comment (Julien00859) [inline addons/sale_stock/models/stock.py:135]: Context leaking

File: addons/stock/models/stock_picking.py [hunk 17]
```diff
@@ -1089,6 +1089,7 @@ def action_see_move_scrap(self):
         action = self.env.ref('stock.action_stock_scrap').read()[0]
         scraps = self.env['stock.scrap'].search([('picking_id', '=', self.id)])
         action['domain'] = [('id', 'in', scraps.ids)]
+        action['context'] = dict(self._context, create=False)
```
Comment (Julien00859) [inline addons/stock/models/stock_picking.py:1092]: Context leaking

File: addons/stock_account/models/stock_inventory.py [hunk 18]
```diff
@@ -31,6 +31,7 @@ def action_get_account_moves(self):
             return False
         action_data = action_ref.read()[0]
         action_data['domain'] = [('stock_move_id.id', 'in', self.move_ids.ids)]
+        action_data['context'] = dict(self._context, create=False)
```
Comment (Julien00859) [inline addons/stock_account/models/stock_inventory.py:34]: Context leaking

**Review by Julien00859 — COMMENTED** [review]
There are bug-reports following the changes done in this branch. See opw-2658775 and attempt fix #79328.

When an action is returned to the client, that this action comes from xml data, then the context that should be used is the context present in the xml data instead of `self.env.context`. Using `self.env.context` leaks a ton of useless information that can be wrongly interpreted by the web-client (see related opw).

Note that the actions coming from xml are string representation of dictionaries that must be evaluated using `safe_eval` and a *carefully crafted* evaluation context. The evaluation context *may* be `self.env.context`.

**Pull request closed without merging**